### DISCUSSION TITLE
Future of "Save Viewer" option

### DISCUSSION THREAD

**Discussion by chylex** [chylex/Discord-History-Tracker#246] [General]
The way DHT viewer currently works, is that Open Viewer / Save Viewer generates a webpage containing the archived data, and then opens it in a browser. This allows the viewer to be used without the app, which can be convenient.

However, this has already somewhat broken down with the option of downloading attachments, which is only supported with Open Viewer, and it works by communicating with the app to fetch the downloaded attachments. If you close the app, the viewer will stop showing attachments altogether. This is not supported at all with Save Viewer - all attachments are fetched through Discord's CDN servers, so any deleted attachments will not work even if they had been downloaded to the archive.

Now that Discord is about to time-limit Discord CDN links, any attachments and embeds that use them will no longer work if you use Save Viewer. You will have to download them to the app (embed downloading is coming soon), and then use Open Viewer to be able to see them. I think that significantly limits the usefulness of Save Viewer.

### The Future of "Open Viewer"

The need to generate webpages with the whole archive has a lot of disadvantages, which could be resolved - but only for Open Viewer. For large databases, it takes a long time to generate the webpage, requires hundreds of MBs of storage per viewer, and even more RAM when opened in a browser.

The radical idea I have is, that the webpage would be served by the DHT app itself, and so would every part of the archived data. The viewer would be lightweight, open instantly, and it would load the channel list, messages, attachments, etc. on-demand by communicating with the app as you browse the archive.

This means the app must be running while you use the viewer, but that should be the only disadvantage; everything else will be massively improved.

### The Future of "Save Viewer"

The improvements to Open Viewer would require most of the viewer to be rewritten, and as I don't have the bandwidth to maintain two separate viewers, Save Viewer would be removed from the app.

With the Discord CDN changes, in order for me to support downloaded attachments etc. in Save Viewer, it would have to export potentially tens or hundreds of GBs of downloaded files next to the viewer, meaning it would either have to diverge from how Open Viewer works and I would end up maintaining two separate viewers anyway, or it would make Open Viewer even worse to use.

I'm not going to do either, so the only remaining options left are 1. removing Save Viewer and making Open Viewer better, or 2. living with the fact that Save Viewer is now worse because Discord CDN links no longer work, and that having to keep the feature around holds back Open Viewer from getting better.

So I would like to hear from you. Do you ever use Save Viewer in the first place, and if so, what use cases do you have that cannot be fulfilled by the improved Open Viewer?

**Comment by Imjustarandomlol32** [comment]
The main reason I use "Save Viewer" is to transfer the html to my phone or for easy sharing, as dht files can be too big to share.

**Reply by chylex** [reply]
Does that change when all attachments and embeds uploaded to Discord will inevitably stop working?

**Reply by Imjustarandomlol32** [reply]
I tend to use it for the messages more than the attachments.
As my cellular data tends to be off, to save internet.

**Reply by chylex** [reply]
I find there's little difference between the dht file and the html viewer once you put it in a zip file, if you don't use the download feature. If you do use the download feature, I've been planning to look into ways of separating the downloaded files from the messages.

**Reply by Imjustarandomlol32** [reply]
Separating like messages on an html and downloads on a zip file?

**Reply by chylex** [reply]
If Save Viewer is removed, there will be no more html. Downloads would be separated into their own dht file so the dht file with messages remains small.

**Reply by Imjustarandomlol32** [reply]
Yeah, that could work. In my opinion, we could have the choice to separate the downloads or not.

**Comment by pyorot** [comment]
Personally, I like to host archives on websites as HTML pages, and Save Viewer is convenient for that. If Discord CDN stops working, then I would upload all the attachments to a different CDN and change the URLs to hotlink them into the HTML page. So for me, downloading all attachments into an adjacent folder and hyperlink replacement with local addresses in the HTML viewer file is ideal.

**Reply by chylex** [reply]
I want to make it possible to host DHT on a server and having it serve the viewer website, but that will require a web hosting that supports .NET. Alternatively, someone else could make a custom exporter that will do what you want, including exporting attachments into a folder. I don't have the bandwidth for it.

**Reply by pyorot** [reply]
Maybe in the spirit of web portability, DHT can be a minified script include built into an html file. Maybe that html file can include a whole DHT database. But I would question whether keeping a whole archive of assets in one file is download-friendly for somebody browsing a viewer website, compared to linking to assets and having them lazily load as you browse.

I've figured out a basic strategy for achieving the kind of archiving I want, which I'll describe here, and it's simple enough to have something like this in a readme or /tools script IMO.

**Reply by pyorot** [reply]
**To create a self-contained archive folder with the viewer html file and an attachments folder:**
1. Generate a Save Viewer html file in DHT as usual, with no attachment downloading needed in DHT.
2. Run a script to list all `cdn.discordapp.com/attachments` URLs from that html in a separate file `i.txt` (slightly bewildered that this required a Python script rather than a simple Bash/PowerShell command involving a custom RegEx).
```python
import re
with open('archive.html', 'r', encoding="utf8") as file:
    text = file.read().replace("\\\\u0026","&")
res = re.findall(r"(https:\/\/cdn.discordapp.com\/attachments\/\d+\/\d+\/[^\?]+(?:[\w\?=]+&){3})", text)
with open('i.txt', "w+", encoding="utf-8") as f:
    f.write("\n".join(res))
```
4. Run Linux/WSL `wget -nv -x -i i.txt` to create the archive (`-nv` = non-verbose output, useful for then finding failed downloads in a text-editor easily; `-x` creates a folder structure mimicking the source; `-i` is input list). I then unnested the attachments folder from the cdn.discordapp.com folder, to be next to the html file.
5. Replace all `https://cdn.discordapp.com/attachments/` URLs in the HTML file with just `attachments/`.
6. Run this bash script in the `attachments` folder to get rid of the query strings in the attachment filenames.
```bash
# /bin/bash
for i in `find $1 -type f`
do
    mv $i `echo $i | cut -d? -f1`
done
```
7. On newer versions of the viewer html file, I had to replace the body of `isImageAttachment(` with this old version to get rid of the URL validation so local addresses can be used.
```js
isImageAttachment(attachment) {
    const extension = attachment.url.split("?")[0].split(".").at(-1);
    return ["png", "gif", "jpg", "jpeg"].includes(extension);
},
```

That's it – the html file now works with a permanent local archive, and it can be zipped up or uploaded to a different CDN with URLs replaced with external ones again.

**Reply by chylex** [reply]
> Maybe in the spirit of web portability, DHT can be a minified script include built into an html file. Maybe that html file can include a whole DHT database.

Well that's exactly what I'm trying to get away from for Open Viewer, so no, the only official solution would be to host the DHT app on a server. Support for this wouldn't be in the first version with the new Open Viewer, though.

**Comment by chylex** [comment]
I made a prototype that has significantly reduced memory usage of both the app and the browser, cut the loading speed in half, and made it possible to load even millions of messages which (at least Firefox) would previously not be able to load at all once the viewer file got past a certain size.

In the current implementation, the app also does not need to continue running once the viewer finishes loading (unless you want to see attachments), but that might change in the future because the current implementation has to pre-load all the messages into the browser and keep them in memory; it just does that in a much more efficient way.

The next DHT update will have the Save Viewer option removed. There might be ways to use the more efficient way of loading messages from a file instead of the app, so Save Viewer could potentially work by saving the data next to the HTML instead of embedding it inside, but at the moment I'm not going to spend time on it.
